### PR TITLE
#64 fix

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
Hi guys!

There is the change list:
- interface IMobileElement was added and AppiumWebElement is implementing it

- the project is referring to WebDriver.Support 2.46.0.0 and Castle.Core 3.3.0.0 now.

- integration with Selenium page object tools. Details are below.
[Information about Page Object design pattern.](https://code.google.com/p/selenium/wiki/PageObjects).  [Information about PageFactory. Here is java version. Sorry. The .Net version works the similar way.](https://code.google.com/p/selenium/wiki/PageFactory)

## The common use case looks as usual.

A single element:
```c#
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]
IWebElement targetElement;
```

or

```c#
using OpenQA.Selenium.Support.PageObjects;
using OpenQA.Selenium.Appium.Interfaces;
using OpenQA.Selenium.Appium;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]
IMobileElement<AppiumWebElement> targetElement;
```

A list of elements:

```c#
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]
IList<IWebElement> targetElements;
```

or

```c#
using OpenQA.Selenium.Support.PageObjects;
using OpenQA.Selenium.Appium.Interfaces;
using OpenQA.Selenium.Appium;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]
IList<AppiumWebElement> targetElement;
```

### The following is also available:

A single element property:

```c#
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]]
private IWebElement TargetElement
{
       set; get;
}
```

or

```c#
using OpenQA.Selenium.Support.PageObjects;
using OpenQA.Selenium.Appium.Interfaces;
using OpenQA.Selenium.Appium;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]]
private IMobileElement<AppiumWebElement> TargetElement
{
       set; get;
}
```

A list of elements property:

```c#
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]]
private IList<IWebElement> TargetElements
{
       set; get;
}
```

or

```c#
using OpenQA.Selenium.Support.PageObjects;
using OpenQA.Selenium.Appium.Interfaces;
using OpenQA.Selenium.Appium;
...

[FindsBy(How = How.ClassName, Using = "theClassOfAnElement")]]
private IList<AppiumWebElement> TargetElement
{
       set; get;
}
```

### How to populate fields and properties

```c#
using OpenQA.Selenium.Appium.PageObjects;
using OpenQA.Selenium.Support.PageObjects;
...

PageObject screenOrPage = new PageObject();
ISearcContext context;
...
PageFactory.InitElements(context, screenOrPage , new AppiumPageObjectMemberDecorator());
```

or 

```c#
using OpenQA.Selenium.Appium.PageObjects;
using OpenQA.Selenium.Support.PageObjects;
...

PageObject screenOrPage = new PageObject();
ISearcContext context;
TimeOutDuration timeSpan = new TimeOutDuration(new TimeSpan(0, 0, 0, 5, 0)); //time 
//of the waiting for elements in general
...
PageFactory.InitElements(context, screenOrPage , new 
AppiumPageObjectMemberDecorator(timeSpan ));
```

### What fields and properties are populated: 

If there is a single element then __IWebElement__ and __IMobileElement<>__ are available. The list of __IMobileElement<>__ parameters is below:
                    - IWebElement
                    - RemoteWebElement
                    - AppiumWebElement
                    - AndroidElement
                    - IOSElement

If there is a list of elements then __IList<>__ with same parameters is can be populated.

The described use case is good when Page/Screen object is for web testing or testing on Android/IOS only. Locators such as __css__, __linkText__ and __partialLinkText__ are invalid for Android/IOS native app testing.

## Advanced use cases

If Page/Screen object is supposed to be reused in crossplatform testing (web testing + Android/iOS native app testing) and a target service/app has the similar  UI for all plaforms then

```c#
using OpenQA.Selenium.Appium.PageObjects.Attributes;
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.Id, Using = "TargetID")] //this locator will be used for Html content at desktop
//or mobile browser/webview
[FindsByIOSUIAutomation(Name = "The name of the iOS native element")] //this locator will be used
//for an iOS native content
[FindsByAndroidUIAutomator(Accesibility = "The accessibility id of the Android native element")] 
//this locator will be used for an Android native content
[FindsBySelendroid(ID = "The selendroid mode element Id")]
IWebElenent target;

//the same is available for lists and element/list properties.
```

### If there are few possible locators

```c#
using OpenQA.Selenium.Appium.PageObjects.Attributes;
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBy(How = How.Id, Using = "TargetID1", Priority = 1)] [FindsBy(How = How.Id, Using = "TargetID2", Priority = 2)] 
[FindsByIOSUIAutomation(Name = "The name of the iOS native element", Priority = 1)] 
[FindsByIOSUIAutomation(Id = "The id of the iOS native element", Priority = 2)] 
//and so on
//The "Priority" property defines a queue of the using of the locator to find elements 
IWebElenent target;

```

### If there is chained searching

```c#
using OpenQA.Selenium.Appium.PageObjects.Attributes;
using OpenQA.Selenium.Support.PageObjects;
...

[FindsBySequence] //is used with Selenium FindsBy
[FindsBy(How = How.Id, Using = "TargetID1", Priority = 1)] [FindsBy(How = How.Id, Using = "TargetID2", Priority = 2)] 
[MobileFindsBySequence(Android = true, IOS = true)] //"Android" and "IOS" properties = true
//mean that the chained searching uses defined locators on these platforms/with these automation
//types
[FindsByIOSUIAutomation(Name = "The name of the iOS native element", Priority = 1)] 
[FindsByIOSUIAutomation(Id = "The id of the iOS native element", Priority = 2)] 
//and so on
IWebElenent target;

```

### If desired elements should match few locators 

then

```c#
using OpenQA.Selenium.Appium.PageObjects.Attributes;
using OpenQA.Selenium.Support.PageObjects;
...

[FindsByAll] //is used with Selenium FindsBy
[FindsBy(How = How.Id, Using = "TargetID1", Priority = 1)] [FindsBy(How = How.ClassName, Using = "TargetID2", Priority = 2)] 
[MobileFindsByAll(Android = true, IOS = true)] //"Android" and "IOS" properties = true
//mean that found elements should match all defined locators on these platforms/with these
// automation types
[FindsByIOSUIAutomation(Name = "The name of the iOS native element", Priority = 1)] 
[FindsByIOSUIAutomation(ID = "The id of the iOS native element", Priority = 2)] 
//and so on
IWebElenent target;

```

### If time of the waiting for elements differs from general

```c#
using OpenQA.Selenium.Appium.PageObjects.Attributes;
using OpenQA.Selenium.Support.PageObjects;
...

[WithTimeSpan(Seconds = 15)]
[FindsBy(How = How.Id, Using = "TargetID")] 
//and so on
IWebElenent target;

```

### If it needs to change general timeout of the waiting for elements 

```c#
using OpenQA.Selenium.Appium.PageObjects;
using OpenQA.Selenium.Support.PageObjects;
...

PageObject screenOrPage = new PageObject();
ISearcContext context;
TimeOutDuration duration = new TimeOutDuration(new TimeSpan(0, 0, 0, 5, 0)); //time 
//of the waiting for elements in general
...
PageFactory.InitElements(context, screenOrPage , new 
AppiumPageObjectMemberDecorator(duration));

//tons of user's code
TimeSpan newTime = new TimeSpan(0, 0, 0, 20, 500);
duration.WaitingDuration = newTime; //and here we go!
```

@Astro03 @Jonahss @bootstraponline 
Please review this. Be careful. There is tons of code : )

PS: Maybe @jimevans would like to look at this PR too ;)

### CODE REVIEW

**Comment by TikhomirovSergey** [comment]
#64

**Comment by bootstraponline** [comment]
I'm not too familiar with .NET so I'll let the experts provide code review. I like the features though. It looks like this will enable the dotnet driver to have the same capabilities as Java in terms of page objects which is great. I think that end users will appreciate having a coherent experience across .NET and Java.

**Comment by Jonahss** [comment]
Great! I'll let the experts do the usual review as well :)

Excellent comments in this pull request, do you want to add them to the README, or add as a doc and link in the readme?

File: appium-dotnet-driver/Appium/PageObjects/AppiumPageObjectMemberDecorator.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            }
+
+            bool hasPropertySet = false;
+            if (property != null)
+            {
+                hasPropertySet = property.CanWrite;
+                targetType = property.PropertyType;
+            }
+
+            if (field == null & (property == null || !hasPropertySet))
```
Comment (Astro03) [inline appium-dotnet-driver/Appium/PageObjects/AppiumPageObjectMemberDecorator.cs]: supposed to be & or &&?

File: appium-dotnet-driver/Appium/PageObjects/Attributes/Abstract/FindsByUIAutomatorsAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OpenQA.Selenium.Appium.PageObjects.Attributes.Abstract
+{
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Class, AllowMultiple = true)]
+    public class FindsByUIAutomatorsAttribute: FindsByMobileAttribute
```
Comment (Astro03) [inline appium-dotnet-driver/Appium/PageObjects/Attributes/Abstract/FindsByUIAutomatorsAttribute.cs]: abstract class?

File: appium-dotnet-driver/Appium/PageObjects/Attributes/Abstract/FindsByMobileAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                throw new ArgumentNullException("obj", "Object to compare cannot be null");
+            }
+
+            FindsByMobileAttribute other = obj as FindsByMobileAttribute;
+            if (other == null)
+            {
+                throw new ArgumentException("Object to compare must be a FindsByAttribute", "obj");
+            }
+
+            // TODO(JimEvans): Construct an algorithm to sort on more than just Priority.
```
Comment (Astro03) [inline appium-dotnet-driver/Appium/PageObjects/Attributes/Abstract/FindsByMobileAttribute.cs]: todo

**Comment by TikhomirovSergey** [comment]
@Jonahss 
I'll start a github wiki. Here will be a page with this description. Links from wiki I'm going to add to the README. What do you think about wiki?

**Comment by Jonahss** [comment]
A wiki for just the .NET driver? Will it include pages which explain the selenium .NET driver too?

**Comment by TikhomirovSergey** [comment]
Yep. It will be wiki for .Net driver.

**Pull request merged**